torch/optim/lr_scheduler.py

```diff
@@ -759,3 +759,220 @@ def step(self, epoch=None):
+            and learning rate is 'base_lr'
+            Default: 0.95
+        div_factor (float): Determines the initial learning rate via
+            initial_lr = max_lr/div_factor
+            Default: 25
+        final_div_factor (float): Determines the minimum learning rate via
+            min_lr = initial_lr/final_div_factor
+            Default: 1e4
+        last_epoch (int): The index of the last batch. This parameter is used when
+            resuming a training job. Since `step()` should be invoked after each
+            batch instead of after each epoch, this number represents the total
+            number of *batches* computed, not the total number of epochs computed.
+            When last_epoch=-1, the schedule is started from the beginning.
+            Default: -1
+
+    Example:
+        >>> data_loader = torch.utils.data.DataLoader(...)
+        >>> optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
+        >>> scheduler = torch.optim.lr_scheduler.OneCycleLR(optimizer, max_lr=0.01, train_dl=data_loader, num_epochs=10)
+        >>> for epoch in range(10):
+        >>>     for batch in data_loader:
+        >>>         train_batch(...)
+        >>>         scheduler.step()
+
+
+    .. _Super-Convergence\: Very Fast Training of Neural Networks Using Large Learning Rates:
+        https://arxiv.org/abs/1708.07120
+    """
+    def __init__(self,
+                 optimizer,
+                 max_lr,
+                 total_steps=None,
+                 num_epochs=None,
+                 train_dl=None,
+                 pct_start=0.3,
+                 anneal_strategy='cos',
+                 cycle_momentum=True,
+                 base_momentum=0.85,
+                 max_momentum=0.95,
+                 div_factor=25.,
+                 final_div_factor=1e4,
+                 last_epoch=-1):
+
+        # Validate optimizer
+        if not isinstance(optimizer, Optimizer):
+            raise TypeError('{} is not an Optimizer'.format(
+                type(optimizer).__name__))
+        self.optimizer = optimizer
+
+        # Validate total_steps
+        if total_steps is None and num_epochs is None and train_dl is None:
+            raise ValueError("You must define either total_steps OR (num_epochs AND train_dl)")
+        elif total_steps is not None:
+            if total_steps <= 0 or not isinstance(total_steps, int):
+                raise ValueError("Expected non-negative integer total_steps, but got {}".format(total_steps))
+            self.total_steps = total_steps
+        else:
+            if num_epochs <= 0 or not isinstance(num_epochs, int):
+                raise ValueError("Expected non-negative integer num_epochs, but got {}".format(num_epochs))
+            if not hasattr(train_dl, '__len__'):
+                raise ValueError("len is not defined over given train_dl")
+            self.total_steps = num_epochs * len(train_dl)
```

Comment:
Is this the only place where data loader is used? If so, would it be better to pass as a parameter the length of the data instead of the loader itself?

Reply:
Yeah, the data loader is only used to help compute the number of total steps in the cycle. Passing only the length of the data loader instead of the data loader itself is an entirely valid approach. I went with this approach because I thought it might arguably be slightly more straightforward for the programmer but I'm certainly not married to the idea whatsoever.

Reply:
Good, let's go with length then. It is more explicit, and I could imagine cases in which length has to be inferred (and so leaving the user in control would be more appropriate).

Reply:
Sounds good. Is the plan for you to more fully review/commandeer this PR once the chainable work is complete?

Reply:
Let's not worry about the other PR (#24352) for now: I can take care of that part :) Besides the comments I left, the PR looks good to me!